### PR TITLE
Fixed issue #120 - unsetting a key can leave keyreport in a bad state

### PR DESCRIPTION
**Pull request by SteveBenz**

### CODE REVIEW

**Comment by NicoHood** [comment]
Any reason why you opened two PRs?
Which one do you prefer and why?

File: src/HID-APIs/DefaultKeyboardAPI.hpp [hunk 1]
```diff
@@ -24,46 +24,56 @@ THE SOFTWARE.
 // Include guard
 #pragma once
 
+DefaultKeyboardAPI::DefaultKeyboardAPI()
+{
+	memset(&_keyReport, 0, sizeof(_keyReport));
```
Comment (NicoHood) [inline src/HID-APIs/DefaultKeyboardAPI.hpp]: As far as I know this is not required, as c++ should init all variable members to zero. Right?

**Comment by SteveBenz** [comment]
As I said in the other review, I think this is the better one.  The other one leaves keystrokes in the same position in the array, and perhaps there's compatibility reasons why that's good.  (Though I don't believe so).  This one is just faster -- in the most common cases, there will be at most one key pressed, and this algorithm will go through the loop at most two times for those cases.  The other implementation will go through the loop 6 times regardless of how many keys are actually pressed.

**Comment by SteveBenz** [comment]
I closed the other review, because I think it's distracting and this is the way to go.  I do have a couple of questions:

1. Making 'set' return true if a key has already been pressed means that we'll end up sending a new report, even when we haven't changed the array at all.  This doesn't seem like a good thing to me.
2. I was testing holding down all 6 available keys, and I found that, even though I know the code is correct in its present form, it appears as if only 5 characters are coming across.  I say that because if I take a plain old keyboard and I 'press asdfjkl' (and hold down the letters as I go), I see it autorepeat on k, the 6th key.  On my keyboard built with HID, it autorepeats on the j, the 5th key.

It's not a big deal, but it's irksome.  I feel like there's a problem somewhere else.

**Comment by SteveBenz** [comment]
(just to be clear, that 5th vs 6th character autorepeat happens the same with or without my changes).  I wonder if there's a descriptor somewhere that's off by one.

**Comment by NicoHood** [comment]
This PR is very confusing to read, but I understand what it does. I have a few questions>
* Compared to the other PR, how much do they differ in size?
* If you reorder the keys on release, isnt this still as same slow as before?
* I think it must be possible to write this code with normal uint8_t i as index, not with pointers. This should make it better to read.

Please open another issue for the 5key thing that you notices. Also please write the OS that you are using and the arduino ide version.

**Comment by SteveBenz** [comment]
**Quick summary** 
The old pull request (my [alternateFix branch](https://github.com/SteveBenz/HID/blob/alternateFix/src/HID-APIs/DefaultKeyboardAPI.hpp)) is 2 just bytes bigger than what's in your master branch right now.  The code in this branch is 16 bytes bigger.  However, a single press-release cycle (the most common case) is two times slower with alternateFix than with master.  Almost all of the win comes from the fact that _press_ always has to loop through the entire array if you allow for holes in the array, but if you don't allow holes, you can always stop at the first zero.  [Converting to arrays](https://github.com/SteveBenz/HID/blob/useIndices/src/HID-APIs/DefaultKeyboardAPI.hpp) significantly increased the code size and slowed performance and I don't think it did anything at all to help readability.

**Dance Mix:**
First, in poking around, I discovered I had an off-by-one error that had a surprisingly large effect on code size and perf.  These numbers reflect the new commit.

Regarding size, I would love to see the assembly, but from what I read, it's a bit of a project to get at it, so I'll settle for the size announced by the compiler.  With the change that we have in this pull request, it grows the code by 16 bytes.  For grins, I swapped out all the pointers for indices (see [line 26](https://github.com/SteveBenz/HID/blob/useIndices/src/HID-APIs/DefaultKeyboardAPI.hpp) and recompiled that.  I got 42 more than the baseline.  So, then I went back to the alternateFix branch (which uses indices, but always loops through the entire array.  To my surprise, that was only 2 bytes more than the baseline.  But hey, given my luck with pointers before, maybe that'd help here?  But nope, after converting it to pointers the code size actually grew by 4 bytes.  Moral of the story: Never try to outguess the C++ compiler.

As to the performance - the difference between master and alternateFix is that alternateFix leaves holes in the list of keys, so _press_ has to iterate through all 6 keys to be correct, always.  Master does not leave those holes, so in both _press_ and _release_ it can stop at the first empty slot.  So, it's natural to expect it to be quicker.  But bigger code is usually slower than smaller code so a test is in order.

I set up [this test](https://github.com/SteveBenz/PS2KeyboardHost/commit/d01287f3feec0618aa07ae7360ebada67af9c3cd).  With "master" it took spewed 9040ms and with the alternateFix it took 15112ms.  Accounting for overhead, it means alternateFix takes ~14us/press-release cycle and with master it's 7.5us/press-release.

As to readability of pointers vs. indices, it really depends a lot on who you are, right?  If you come from an stl background, where iterators are essentially pointers, then I think you'll be more comfortable with pointers.  There are legends that pointers are "hard".  I honestly believe that legend has its roots in Fortran, where there were no pointers.  CompSci professors of old weren't comfortable using them, and that carried on to their students & students' students.  But today, I work with many recent grads, and they just don't struggle with it.  (Although I should say that the company I work for can afford to be very picky.)

If you are really serious about readability, I'd suggest the way to go would be to split 'set' into separate press/release methods.  That would make a big difference in readability at a modest cost to code size.  Another good opportunity would be to meld DefaultKeyboard with DefaultKeyboardAPI and lose the fancy 'using' statements in DefaultKeyboardAPI.

**Comment by NicoHood** [comment]
Closing as #148 fixed it already

**Pull request closed without merging**